### PR TITLE
ShyLU_node, FastILU : fixing performance issue

### PR DESCRIPTION
**Pull request by iyamazaki**
@trilinos/shylu 

## Motivation

This PR contains changes for fixing performance of FastILU

- changes to improves the compute time
- option to run SpTRSV on host for solve

## Testing

Modify the tester to add the change.

## Additional Information

The same changes should be made for FastILDL and FastIC

### CODE REVIEW

**Review by jhux2 — CHANGES_REQUESTED** [review]
@iyamazaki Thank you for this enhancement!  Could you please add these options to the Ifpack2 user's manual?

**Comment by iyamazaki** [comment]
@jhux2 That is a good idea. Can you tell me how I do that?

**Comment by jhux2** [comment]
> @jhux2 That is a good idea. Can you tell me how I do that?

The LaTeX source is in `ifpack2/doc/UsersGuide`.

**Comment by iyamazaki** [comment]
Thank you, @jhux2. I'll update the document.

By the way, I update ctest to include new options, and locally test (help from @vqd8a). But, I don't think this solver is enabled in any regular tests (PR or nightly)?

**Comment by jhux2** [comment]
> By the way, I update ctest to include new options, and locally test (help from @vqd8a). But, I don't think this solver is enabled in any regular tests (PR or nightly)?

I'm not sure.  But it's fairly straightforward to add specialized nightly tests.  MueLu and Tpetra already have their own tests in the Experimental track.

File: packages/shylu/shylu_node/fastilu/src/shylu_fastilu.hpp [hunk 1]
```diff
@@ -482,47 +506,75 @@ class FastILUPrec
...
         void numericILU()
         {
-            aVal = ScalarArray("aVal", aColIdx.extent(0));
-            aVal_ = Kokkos::create_mirror(aVal);
+            const Scalar zero = STS::zero();
+            #ifdef FASTILU_TIMER
+            Kokkos::Timer Timer;
+            #endif
+            //Sort each row of A by ColIdx
+            KokkosKernels::sort_crs_matrix<ExecSpace, OrdinalArray, OrdinalArray, ScalarArray>(aRowMapIn, aColIdxIn, aValIn);
```
Comment (srajama1) [inline packages/shylu/shylu_node/fastilu/src/shylu_fastilu.hpp:522]: Can this be in symbolic?

Comment (iyamazaki) [reply packages/shylu/shylu_node/fastilu/src/shylu_fastilu.hpp:522]: @srajama1, we internally sort the "graph" in symbolic, but when the user inputs the "matrix" for numeric, the matrix is not sorted? We could insert the numerical value, looking for ColIdx match, but sorting the matrix seems to go faster.

Comment (srajama1) [reply packages/shylu/shylu_node/fastilu/src/shylu_fastilu.hpp:522]: Understood. Thanks! Typically the structure should not change for numeric, so the position should be the same, but this is ok.

**Review by vqd8a — APPROVED** [review]
The changes look good to me. Thanks @iyamazaki for fixing the performance issue. Tested with Aria matrices. Compute time is significantly faster with these changes.

**Comment by vqd8a** [comment]
@srajama1 Is this okay to merge?

**Comment by iyamazaki** [comment]
Thank you everyone for the reviews!! Just a note for @srajama1, I meant: we internally store the sorted graph and matrix, but our current symbolic doesn't save the "mapping" from the non-sorted input ColIdx to the internal sorted ColIdx. That is why we are sorting again the input matrix in numeric. Not sure if it is clearer but if sort-time becomes significant, I'll look into storing the "map" in symbolic.

**Comment by srajama1** [comment]
@iyamazaki That is reasonable. I will approve it.

**Review by srajama1 — APPROVED** [review]

**Pull request merged**